### DISCUSSION TITLE
RFC: Offline docs

### DISCUSSION THREAD

**Discussion by BrunoQuaresma** [coder/coder#8110] [RFCs]
We have a few users that need to run Coder from air-gapped networks, and to make that work fully, we need to support offline docs.

## Constraints
- We don't want to increase the bundle size significantly. Right now, embedding the docs directly would add ~11MB to the binary which is not desired. 

## Proposed solution
- Pack the coder docs into a separate package/artifact during release so the operator could host the docs markdown files locally or anywhere else.
- Add an option to the Coder CLI to set the path or URL to the custom docs path `coder server --offline-docs <docs-path>` that could be a URL or a path to load the markdown files. It could also be an env variable like `OFFLINE_DOCS=<docs-path> coder server`. 
- Add the `offline_docs` config into the deployment endpoint so the FE can knows it is using or not offline docs. It is also used from the FE to fetch the markdown files content.
- Add into the FE the related docs pages.
- Change the docs links in the app to use or not the offline docs if the options is enabled.

## Possible issues
- The docs files need to be hosted in a web server so the FE can use the images and fetch the content properly. One possible solution would be to move the content fetching into the BE but we still need to have access to the images so it would not remove the web server dependency entirely.
- The docs should be hosted under the same domain to avoid CORS issues.

**Comment by spikecurtis** [comment]
A quick & simple solution would be to produce a docs container image, or just a tar of the webroot to drop into NGINX, Apache, or whatever.  We don't need to modify the dashboard.  The operator can just host the docs wherever makes sense in their network.

If we get significant customer interest in tighter docs integration we could work on the above proposal, but I think we should stick to something simple at first.

**Reply by mafredri** [reply]
I like it, then we would only need to add support for docs base url in the product and we can assumed it can be reached.

**Reply by BrunoQuaresma** [reply]
One issue is, we would have to parse the markdown and navigation to present it to the user somehow. I think this is not very easy because our docs schema is very focused on our use case and written by us. So probably, creating an image with docs + some markdown docs parser + adapting our docs to fit this parser would be more difficult.

**Reply by deansheather** [reply]
This could be rendered at release time into static HTML + CSS (with JS only for search or something)

**Reply by BrunoQuaresma** [reply]
I mean, if we need to render it into a release time into a static HTML + CSS - we probably would need to do it ourselves - I don't see too much work difference by adding it into the UI directly tho.

**Reply by spikecurtis** [reply]
How are the docs rendered today?  Can't we just use the same tooling to render a static HTML + CSS output?

**Reply by ammario** [reply]
👍🏽 to hosting a static site from a Docker container, if our docs code is agreeable.

It would be great to avoid adding complexity to our core product and the coder.com codebase.

**Reply by BrunoQuaresma** [reply]
@spikecurtis today, the docs tooling lives inside of the website codebase using NextJS. But the tooling for it is kinda unique because it handles multiple versions, multiple products and searching. We can extract this of course, I'm just finding it more complex to do than add it into the UI.

**Reply by BrunoQuaresma** [reply]
I think the proposed solution from @spikecurtis sounds the best so far. In my mind, the steps to achieve that would be:
- Create a docs website that can parse the markdown, parse the navigation, and style it correctly. Preferably generates static pages.
- Pack this docs website into a docker image or a simple .tar file
- Create a docs artifact on every release together with the coder binary

@spikecurtis does it sound accurate?

**Reply by ammario** [reply]
@BrunoQuaresma — in your vision does our official, live docs at https://docs.coder.com/v2/ reuse the static site code? It would be nice to avoid significant divergence since it's easy to forget about (and regress) offline docs when such a small portion of our userbase uses it.

**Reply by BrunoQuaresma** [reply]
It is kinda hard because of the engine that we use for the docs on the coder.com handle multiple products and versions. It also fetches data from GitHub instead of local files to support multiple versions. I would be more concerned about divergence in the content which is not going to be the case tho since they would use the same "source". Maybe, when we deprecate v1 and don't need to support multiple products or versions, we could use the same codebase for them without too much extra work. 

Does it answer the question? Not sure if I understood it correctly.

**Reply by ammario** [reply]
Sort of. Keeping the content the same was a must for me. Even if we separated v1, we have code-server docs.

Curious to hear more about how you would change the current codebase to support static output.

**Reply by BrunoQuaresma** [reply]
> Keeping the content the same was a must for me.

Yes, and they will be of course.

> Even if we separated v1, we have code-server docs

Hm.. yeah, you are right.

> Curious to hear more about how you would change the current codebase to support static output.

What current codebase? coder.com?

**Reply by ammario** [reply]
Yeah, coder.com

**Reply by BrunoQuaresma** [reply]
I don't think it is important to have offline docs and the website looking the same because they have different purposes. This looks an early abstraction thing when they look similar but have different context

**Reply by ammario** [reply]
I agree they don't have to look the same. My desire for code reuse is based on a worry that we may break the Offline Docs features without noticing it because of the sparse usage.

**Reply by BrunoQuaresma** [reply]
Since it would be generated statically we would get the errors during the build tho

**Reply by BrunoQuaresma** [reply]
At least the "broken" ones. Things like wrong links would not be catch but we can do some testing to avoid this

**Comment by BrunoQuaresma** [comment]
So the final decision is:
- Copy the docs engine we have from offline docs in v1
- Build it during the release
- Add docs about this feature
- Add tests to catch any missed links in the docs
